<div style="text-align: center;">
<h1> Zomato Dataset Analysis</h1>
</div>

## Importing the required libraries and loading the dataset

In [163]:
# import the libraries
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
import seaborn as sns
import re

In [164]:
df = pd.read_csv(r"..\Datasets\zomato.csv")

In [165]:
print(df)

                                                 address  \
0      942, 21st Main Road, 2nd Stage, Banashankari, ...   
1      2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...   
2      1112, Next to KIMS Medical College, 17th Cross...   
3      1st Floor, Annakuteera, 3rd Stage, Banashankar...   
4      10, 3rd Floor, Lakshmi Associates, Gandhi Baza...   
...                                                  ...   
56247  Four Points by Sheraton Bengaluru, 43/3, White...   
56248  Number 10, Garudachar Palya, Mahadevapura, Whi...   
56249  Sheraton Grand Bengaluru Whitefield Hotel & Co...   
56250  Sheraton Grand Bengaluru Whitefield Hotel & Co...   
56251  ITPL Main Road, KIADB Export Promotion Industr...   

                                                    name online_order  \
0                                                  Jalsa          Yes   
1                                         Spice Elephant          Yes   
2                                        San Churro Cafe    

## Understanding the data

In [166]:
# checking all the column details
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56252 entries, 0 to 56251
Data columns (total 13 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   address                      56235 non-null  object
 1   name                         56236 non-null  object
 2   online_order                 56233 non-null  object
 3   book_table                   56194 non-null  object
 4   rate                         48414 non-null  object
 5   votes                        56174 non-null  object
 6   phone                        54956 non-null  object
 7   location                     56126 non-null  object
 8   rest_type                    55914 non-null  object
 9   dish_liked                   28027 non-null  object
 10  cuisines                     56049 non-null  object
 11  approx_cost(for two people)  55731 non-null  object
 12  listed_in(type)              51642 non-null  object
dtypes: object(13)
memory usage: 5.6

In [167]:
# no. of rows and columns
df.shape

(56252, 13)

In [168]:
# data types
df.dtypes

address                        object
name                           object
online_order                   object
book_table                     object
rate                           object
votes                          object
phone                          object
location                       object
rest_type                      object
dish_liked                     object
cuisines                       object
approx_cost(for two people)    object
listed_in(type)                object
dtype: object

In [169]:
# top rows of the dataset
df.head()

address                   name  \
0  942, 21st Main Road, 2nd Stage, Banashankari, ...                  Jalsa   
1  2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...         Spice Elephant   
2  1112, Next to KIMS Medical College, 17th Cross...        San Churro Cafe   
3  1st Floor, Annakuteera, 3rd Stage, Banashankar...  Addhuri Udupi Bhojana   
4  10, 3rd Floor, Lakshmi Associates, Gandhi Baza...          Grand Village   

  online_order book_table   rate votes                             phone  \
0          Yes        Yes  4.1/5   775    080 42297555\r\n+91 9743772233   
1          Yes         No  4.1/5   787                      080 41714161   
2          Yes         No  3.8/5   918                    +91 9663487993   
3           No         No  3.7/5    88                    +91 9620009302   
4           No         No  3.8/5   166  +91 8026612447\r\n+91 9901210005   

       location            rest_type  \
0  Banashankari        Casual Dining   
1  Banashankari        Casual Dining   
2  Banashankari  Cafe, Casual Dining   
3  Banashankari          Quick Bites   
4  Basavanagudi        Casual Dining   

                                          dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         cuisines approx_cost(for two people) listed_in(type)  
0  North Indian, Mughlai, Chinese                         800          Buffet  
1     Chinese, North Indian, Thai                         800          Buffet  
2          Cafe, Mexican, Italian                         800          Buffet  
3      South Indian, North Indian                         300          Buffet  
4        North Indian, Rajasthani                         600          Buffet

In [170]:
# botttom column
df.tail()

address  \
56247  Four Points by Sheraton Bengaluru, 43/3, White...   
56248  Number 10, Garudachar Palya, Mahadevapura, Whi...   
56249  Sheraton Grand Bengaluru Whitefield Hotel & Co...   
56250  Sheraton Grand Bengaluru Whitefield Hotel & Co...   
56251  ITPL Main Road, KIADB Export Promotion Industr...   

                                                    name online_order  \
56247  Best Brews - Four Points by Sheraton Bengaluru...           No   
56248                           Vinod Bar And Restaurant           No   
56249  Plunge - Sheraton Grand Bengaluru Whitefield H...           No   
56250  Chime - Sheraton Grand Bengaluru Whitefield Ho...           No   
56251                       The Nest - The Den Bengaluru           No   

      book_table    rate votes           phone                    location  \
56247         No  3.6 /5    27    080 40301477                  Whitefield   
56248         No     NaN     0  +91 8197675843                  Whitefield   
56249         No     NaN     0             NaN                  Whitefield   
56250        Yes  4.3 /5   236    080 49652769  ITPL Main Road, Whitefield   
56251         No  3.4 /5    13  +91 8071117272  ITPL Main Road, Whitefield   

                rest_type                    dish_liked  \
56247                 Bar                           NaN   
56248                 Bar                           NaN   
56249                 Bar                           NaN   
56250                 Bar  Cocktails, Pizza, Buttermilk   
56251  Bar, Casual Dining                           NaN   

                                     cuisines approx_cost(for two people)  \
56247                             Continental                       1,500   
56248                             Finger Food                         600   
56249                             Finger Food                       2,000   
56250                             Finger Food                       2,500   
56251  Finger Food, North Indian, Continental                       1,500   

      listed_in(type)  
56247   Pubs and bars  
56248   Pubs and bars  
56249   Pubs and bars  
56250   Pubs and bars  
56251   Pubs and bars

In [171]:
# random rows
df.sample(5)

address  \
11959  93, 17th Main Road, Koramangala 5th Block, Ban...   
9786   99, 100ft Ring Road, 2nd Colony, 1st Stage, BT...   
8404   Bengaluru Marriott Hotel, 75, 8th Road, EPIP A...   
45772  Ground Floor, 14, HM Geneva House, Cunningham ...   
50998   I'd have the course meal with just soups. It ...   

                                                    name  \
11959                                     Malabar Xpress   
9786                                   Tasty Jigarthanda   
8404     Alto Vino - Bengaluru Marriott Hotel Whitefield   
45772                                                KFC   
50998   Sushis for main course and Lamingtons slice f...   

                                            online_order     book_table  \
11959                                                Yes             No   
9786                                                 Yes             No   
8404                                                  No            Yes   
45772                                                Yes             No   
50998   right from the start. From explaining their s...   ('Rated 4.0'   

                                                    rate  \
11959                                                NaN   
9786                                               3.4/5   
8404                                               4.3/5   
45772                                             3.6 /5   
50998   ""RATED\n  Nicely done place though I would s...   

                                                   votes  \
11959                                                  0   
9786                                                  10   
8404                                                 263   
45772                                                211   
50998   they have quite a nice spread. Here's what I ...   

                                                   phone  \
11959                   +91 8095556677\r\n+91 8095556688   
9786                                      +91 9790447882   
8404                                        080 49653131   
45772                                       080 33994444   
50998   not good. The vegetables were stringy!\nTeriy...   

                                                location  \
11959                              Koramangala 5th Block   
9786                                                 BTM   
8404                                          Whitefield   
45772                                    Cunningham Road   
50998   have had better..\nSchezuan dim sum - excelle...   

                                               rest_type  \
11959                                        Quick Bites   
9786                                       Beverage Shop   
8404                                         Fine Dining   
45772                                        Quick Bites   
50998   a lot of thought has gone into creating the v...   

                                              dish_liked  \
11959                                                NaN   
9786                                                 NaN   
8404   Tiramisu, Wine, Lamb, Baked Salmon, Chicken Pa...   
45772  Burgers, Crispy Chicken, Chicken Zinger Burger...   
50998   hats off for that. Will definitely visit to t...   

                   cuisines  \
11959  South Indian, Kerala   
9786   Beverages, Ice Cream   
8404         Italian, Salad   
45772     Burger, Fast Food   
50998          ('Rated 5.0'   

                             approx_cost(for two people)  \
11959                                                300   
9786                                                 150   
8404                                               3,000   
45772                                                400   
50998   'RATED\n  Food : 5/5 || Ambience : 4/5 || Val...   

                                         listed_in(type)  
11959                                           Dine-out  
9786                                        

### Conclusion
- There are 12 columns in the dataset
- 56252 rows in the dataset
- All the colums data types are same

## Data Cleaning

In [172]:
df.head(1)

address   name online_order  \
0  942, 21st Main Road, 2nd Stage, Banashankari, ...  Jalsa          Yes   

  book_table   rate votes                           phone      location  \
0        Yes  4.1/5   775  080 42297555\r\n+91 9743772233  Banashankari   

       rest_type                                         dish_liked  \
0  Casual Dining  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   

                         cuisines approx_cost(for two people) listed_in(type)  
0  North Indian, Mughlai, Chinese                         800          Buffet

**Checking the duplicate values and dropping them**

In [190]:
df.duplicated().sum()

np.int64(17943)

In [191]:
df.drop_duplicates(inplace=True)

In [192]:
df.duplicated().sum()

np.int64(0)

In [194]:
df.shape

(38309, 13)

**Standardize column names**

In [173]:
df.head(1)

address   name online_order  \
0  942, 21st Main Road, 2nd Stage, Banashankari, ...  Jalsa          Yes   

  book_table   rate votes                           phone      location  \
0        Yes  4.1/5   775  080 42297555\r\n+91 9743772233  Banashankari   

       rest_type                                         dish_liked  \
0  Casual Dining  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   

                         cuisines approx_cost(for two people) listed_in(type)  
0  North Indian, Mughlai, Chinese                         800          Buffet

In [174]:
df.columns

Index(['address', 'name', 'online_order', 'book_table', 'rate', 'votes',
       'phone', 'location', 'rest_type', 'dish_liked', 'cuisines',
       'approx_cost(for two people)', 'listed_in(type)'],
      dtype='object')

In [175]:
df.columns = [c.strip().lower().replace(' ', '_').replace('(', '').replace(')', '') for c in df.columns]

In [176]:
df = df.rename(columns={
    'approx_costfor_two_people': 'approx_cost_for_two'
})

In [177]:
df.columns

Index(['address', 'name', 'online_order', 'book_table', 'rate', 'votes',
       'phone', 'location', 'rest_type', 'dish_liked', 'cuisines',
       'approx_cost_for_two', 'listed_intype'],
      dtype='object')

In [178]:
df.head(1)

address   name online_order  \
0  942, 21st Main Road, 2nd Stage, Banashankari, ...  Jalsa          Yes   

  book_table   rate votes                           phone      location  \
0        Yes  4.1/5   775  080 42297555\r\n+91 9743772233  Banashankari   

       rest_type                                         dish_liked  \
0  Casual Dining  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   

                         cuisines approx_cost_for_two listed_intype  
0  North Indian, Mughlai, Chinese                 800        Buffet

**parse rating column -> numeric (4.1/5 -> 4.1, NEW/- -> NaN)**

In [179]:
def parse_rate(x):
    x = str(x).strip() # Convert to string & remove spaces.

    if x.upper() in {'NEW','-','--',''}:    # If the value is NEW, -, -- or empty → return NaN.
        return np.nan

    if '/' in x: # If rating looks like '4.1/5'
        try:
            return float(x.split('/')[0].strip())  # Extract '4.1' and convert to float.
        except:
            return np.nan

    try:
        return float(x) # Try converting normally (e.g. '4.3').
    except:
        return np.nan # If conversion fails → return NaN.


In [180]:
# applying on the rate column
df['rate'].apply(parse_rate)

0        4.1
1        4.1
2        3.8
3        3.7
4        3.8
        ... 
56247    3.6
56248    NaN
56249    NaN
56250    4.3
56251    3.4
Name: rate, Length: 56252, dtype: float64

In [181]:
# re-assigning in the column
df['rate'] = df['rate'].apply(parse_rate)

In [182]:
df.sample(2)

address  \
46111  3137/A, 7th Main, HAL 2nd Stage, Off Double Ro...   
54503    24-26, 414/80/2, Varthur Hobli, Outer Ring Road   

                                name online_order book_table  rate votes  \
46111                Ooty Chocolates          Yes         No   3.9    55   
54503  Tipsy Bull - The Bar Exchange          Yes        Yes   4.4  2232   

                              phone     location           rest_type  \
46111  +91 9845819592\n080 41154280  Indiranagar      Dessert Parlor   
54503                  080 49653133    Bellandur  Bar, Casual Dining   

                                              dish_liked  \
46111                                                NaN   
54503  Falafel, Cocktails, Mocktails, Sea Food, Sprin...   

                                          cuisines approx_cost_for_two  \
46111                                     Desserts                 300   
54503  North Indian, Chinese, Continental, Mexican               1,400   

       listed_intype  
46111       Delivery  
54503  Pubs and bars

**phone -> keep first phone, digits and leading**

In [183]:
def clean_phone(x):
    # If the value is NaN, return NaN immediately
    if pd.isna(x): 
        return np.nan
    
    # Convert to string and split at '\r\n' or '\n' to keep only the first phone number
    s = str(x).split('\\r\\n')[0].split('\n')[0]
    
    # Remove all characters except digits and '+'
    s = re.sub(r'[^0-9+]', '', s)
    
    # If the final cleaned string is empty, return NaN; otherwise return the cleaned phone
    return s if s else np.nan


In [184]:
df['phone'].apply(clean_phone)

0          08042297555
1          08041714161
2        +919663487993
3        +919620009302
4        +918026612447
             ...      
56247      08040301477
56248    +918197675843
56249              NaN
56250      08049652769
56251    +918071117272
Name: phone, Length: 56252, dtype: object

**Converting to proper data type**

In [185]:
df.head(1)

address   name online_order  \
0  942, 21st Main Road, 2nd Stage, Banashankari, ...  Jalsa          Yes   

  book_table  rate votes                           phone      location  \
0        Yes   4.1   775  080 42297555\r\n+91 9743772233  Banashankari   

       rest_type                                         dish_liked  \
0  Casual Dining  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   

                         cuisines approx_cost_for_two listed_intype  
0  North Indian, Mughlai, Chinese                 800        Buffet

In [186]:
df.dtypes   

address                 object
name                    object
online_order            object
book_table              object
rate                   float64
votes                   object
phone                   object
location                object
rest_type               object
dish_liked              object
cuisines                object
approx_cost_for_two     object
listed_intype           object
dtype: object

In [187]:
# convert yes/no to boolean
df["online_order"] = df["online_order"].map({"Yes": True, "No": False})
df["book_table"] = df["book_table"].map({"Yes": True, "No": False})

In [188]:
df.sample(2)

address                name  \
4401  Ground Floor, Ganapa Tower, Outer Ring Road, B...  Sri Murali Krishna   
9040   Near McDonald's, 4th Block, Jayanagar, Bangalore          SL Vadapav   

     online_order book_table  rate votes           phone   location  \
4401         True      False   NaN     0  +91 7676094562  Bellandur   
9040         True      False   4.0    79  +91 9900597313  Jayanagar   

          rest_type                                   dish_liked  \
4401  Casual Dining                                          NaN   
9040    Quick Bites  Dabeli, Samosa, Tawa Pulav, Cheese Vada Pav   

                            cuisines approx_cost_for_two listed_intype  
4401  North Indian, Chinese, Biryani                 600      Delivery  
9040                   Maharashtrian                 100      Delivery